gold/nanomips.cc

```diff
@@ -5598,6 +5753,79 @@ Nanomips_expand_insn_finalize<size, big_endian>::type(
   return type;
 }
 
+// Nanomips_trampoline methods.
+
+// Return matching BALC instruction for trampoline if there is one.
+
+template<int size, bool big_endian>
+const Nanomips_insn_property*
+Nanomips_trampoline<size, big_endian>::find_insn(
+    Nanomips_relobj<size, big_endian>*,
+    uint32_t insn,
+    unsigned int mask,
+    unsigned int r_type)
+{
+  switch (r_type)
+    {
+    case elfcpp::R_NANOMIPS_PC25_S1:
+      return nanomips_insn_property_table->get_insn_property(insn, mask,
+                                                             r_type);
+    default:
+      break;
+    }
+  return NULL;
+}
+
+// Return the transformation type if instruction needs to be expanded.
+
+template<int size, bool big_endian>
+unsigned int
+Nanomips_trampoline<size, big_endian>::type(
+    const Relocate_info<size, big_endian>* relinfo,
+    Target_nanomips<size, big_endian>* target,
+    const Symbol*,
+    const Symbol_value<size>* psymval,
+    const Nanomips_insn_property* insn_property,
+    const elfcpp::Rela<size, big_endian>& reloc,
+    size_t,
+    uint32_t,
+    Address address,
+    Address,
+    bool has_balc_stub2)
+{
+  unsigned int r_type = elfcpp::elf_r_type<size>(reloc.get_r_info());
+
+  if ((r_type != elfcpp::R_NANOMIPS_PC25_S1 || !has_balc_stub2)
```

Comment:
Unnecessary nesting. Would be better to put all parts of the || expression at the same level.
I am inclined to think that the check for the instruction should be an assertion. If you received a NOTRAMP relocation.

Also note the comment about the name `has_balc_stub2' below. I'd expect has_balc_stub2 to imply that it is a 32-bit call instruction and already has the appropriate relocation and any checking to that effect would simply be in an assert( ).